### PR TITLE
add multi-field scoping for slugs

### PR DESCRIPTION
**Pull request by mikekosulin**
* Add option for multi-field scoping in slugs, expanding beyond single-field/relation scopes without altering existing functionality

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,8 @@
-## 7.0.1 (Next)
+## 7.1.1 (Next)
 
-* Your contribution here.
+## 7.1.0 (2023/10/20)
```
Comment (dblock) [inline CHANGELOG.md:3]: We haven't released 7.1.0. Let's increment the version to 7.1.0 in this PR, it should say `7.1.0 (Next)`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,8 @@
...
+
+* [#274](https://github.com/mongoid/mongoid-slug/pull/274): Added support for scoping slugs by multiple fields
```
Comment (dblock) [inline CHANGELOG.md:5]: Make this like the rest of the lines with your GitHub username.

File: lib/mongoid/slug.rb [hunk 3]
```diff
@@ -90,8 +91,18 @@ def slug(*fields, &block)
 
         # Set indexes
         if slug_index && !embedded?
-          Mongoid::Slug::IndexBuilder.build_indexes(self, slug_scope_key, slug_by_model_type,
-                                                    options[:localize])
+          # Check if scope is an array and handle accordingly.
+          if slug_scope.is_a?(Array)
```
Comment (dblock) [inline lib/mongoid/slug.rb:95]: Probably simpler to do `Array(slug_scope).each do |...|`, avoids code duplication.

File: lib/mongoid/slug.rb [hunk 4]
```diff
@@ -119,7 +130,14 @@ def look_like_slugs?(*args)
       def slug_scope_key
         return nil unless slug_scope
 
-        reflect_on_association(slug_scope).try(:key) || slug_scope
+        # If slug_scope is an array, we return an array of keys.
+        if slug_scope.is_a?(Array)
```
Comment (dblock) [inline lib/mongoid/slug.rb:134]: Same as above, `Array(slug_scope).each`

File: lib/mongoid/slug/unique_slug.rb [hunk 5]
```diff
@@ -100,10 +100,15 @@ def find_unique(attempt = nil)
           where_hash[:_slugs.all] = [regex_for_slug]
           where_hash[:_id.ne]     = model._id
 
-          if (scope = slug_scope) && reflect_on_association(scope).nil?
-            # scope is not an association, so it's scoped to a local field
-            # (e.g. an association id in a denormalized db design)
-            where_hash[scope] = model.try(:read_attribute, scope)
+          if (scope = slug_scope)
+            scopes = scope.is_a?(Array) ? scope : [scope] # Wrap single scope into an array for uniform handling
+            scopes.each do |individual_scope|
```
Comment (dblock) [inline lib/mongoid/slug/unique_slug.rb:105]: `Array(slug_scope).each ...`

**Review by dblock — COMMENTED** [review]
Looks great! Let's simplify the code a bit with a uniform `Array(...).each ...`

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,6 @@
-## 7.0.1 (Next)
+## 7.1.0 (Next)
```
Comment (dblock) [inline CHANGELOG.md:1]: Increment this in https://github.com/mongoid/mongoid-slug/blob/master/lib/mongoid/slug/version.rb as well please.

File: README.md [hunk 7]
```diff
@@ -193,6 +193,26 @@ class Employee
...
+  field :name
+  field :company_id
+  field :department_id
+
+  # Scope slug uniqueness by a combination of company and department
+  slug :name, scope: %i[company_id department_id]
+end
+```
+
+Note: This approach creates multiple indexes, differing from single-field scoping, and impacting database performance and storage.
```
Comment (dblock) [inline README.md:214]: Nit: lowercase 'This' :)

File: lib/mongoid/slug.rb [hunk 8]
```diff
@@ -90,8 +91,14 @@ def slug(*fields, &block)
 
         # Set indexes
         if slug_index && !embedded?
-          Mongoid::Slug::IndexBuilder.build_indexes(self, slug_scope_key, slug_by_model_type,
-                                                    options[:localize])
+          # Even if the slug_scope is nil, we need to proceed.
+          slug_scopes = slug_scope.nil? ? [slug_scope_key] : Array(slug_scope)
```
Comment (dblock) [inline lib/mongoid/slug.rb:95]: The implementation of `slug_scope_key` iterates over slug scopes that can be `nil`, then returns `nil` in the case of `slug_scope` being nil.  

I suggest adding a `slug_scopes` method that does not call `slug_scope_key`:

```ruby
def slug_scopes
    slug_scope ? Array(slug_scope) : [nil]
end
```

File: lib/mongoid/slug/unique_slug.rb [hunk 9]
```diff
@@ -100,10 +100,14 @@ def find_unique(attempt = nil)
           where_hash[:_slugs.all] = [regex_for_slug]
           where_hash[:_id.ne]     = model._id
 
-          if (scope = slug_scope) && reflect_on_association(scope).nil?
-            # scope is not an association, so it's scoped to a local field
-            # (e.g. an association id in a denormalized db design)
-            where_hash[scope] = model.try(:read_attribute, scope)
+          if (scope = slug_scope)
```
Comment (dblock) [inline lib/mongoid/slug/unique_slug.rb:103]: Is the `scope` assignment unnecessary here? 

```
Array(slug_scope).each ... 
```

**Review by dblock — CHANGES_REQUESTED** [review]
Much better, but the implementation is still a little confusing. See below. Thanks for hanging in here with me!

File: lib/mongoid/slug.rb (hunk 8 above)
Comment (mikekosulin) [reply lib/mongoid/slug.rb:95]: Got it.

Also I believe there's no need to change the way we invoke the index builder and we should keep it as it was in version 7.

```ruby
if slug_index && !embedded?
  Mongoid::Slug::IndexBuilder.build_indexes(self, slug_scope_key, slug_by_model_type,
                                            options[:localize])
end
```

But we've to modify the [index_builder.rb](https://github.com/mongoid/mongoid-slug/blob/6b6ad2f07f4772f5615dc8492c3ba870afb017c5/lib/mongoid/slug/index_builder.rb#L31) this way:

```diff
- fields[scope_key] = 1 if scope_key
+ Array(scope_key).each { |key| fields[key] = 1 }
```

Comment (dblock) [reply lib/mongoid/slug.rb]: As long as all test pass ;)

File: lib/mongoid/slug.rb [hunk 10]
```diff
@@ -113,13 +113,23 @@ def look_like_slugs?(*args)
...
+      def slug_scopes
+        # If slug_scope is set (i.e., not nil), we convert it to an array to ensure we can handle it consistently.
+        # If it's not set, we use an array with a single nil element, signifying no specific scope.
+        slug_scope ? Array(slug_scope) : [nil]
+      end
+
       # Returns the scope key for indexing, considering associations
       #
       # @return [ Array<Document>, Document ]
       def slug_scope_key
```
Comment (dblock) [inline lib/mongoid/slug.rb:125]: Should this be renamed to `slug_scope_keys`?

**Review by dblock — COMMENTED** [review]
I think version.rb is the last thing.

**Comment by dblock** [comment]
@johnnyshields any comments/issues?

File: lib/mongoid/slug.rb (hunk 10 above)
Comment (mikekosulin) [reply lib/mongoid/slug.rb:125]: Makes sense

**Review by dblock — APPROVED** [review]
LGTM! Thanks. I'll wait for @johnnyshields to chime in, will merge this week if we don't hear from him.

**Comment by mikekosulin** [comment]
Hey @dblock, @johnnyshields any updates on merging? 😊

File: README.md [hunk 11]
```diff
@@ -193,6 +193,26 @@ class Employee
...
+  field :name
+  field :company_id
+  field :department_id
+
+  # Scope slug uniqueness by a combination of company and department
+  slug :name, scope: %i[company_id department_id]
+end
+```
+
+Note: this approach creates multiple indexes, differing from single-field scoping, and impacting database performance and storage.
```
Comment (johnnyshields) [inline README.md:214]: Hmmm... shouldn't this still create only one _compound_ index? For example:

```
slug :name, scope: %i[company_id department_id]

# Should create
index company_id: 1, department_id: 1, _slugs: 1

# NOT
index company_id: 1, _slugs: 1
index department_id: 1, _slugs: 1
```

Comment (mikekosulin) [reply README.md]: Sorry for the late response. You're right, it creates only a compound index in this case. I've [fixed it here](https://github.com/mongoid/mongoid-slug/pull/274/commits/601318844bdaf81212754d4212868fb28aca8208) and forgot to remove the note.

File: lib/mongoid/slug.rb [hunk 12]
```diff
@@ -297,7 +307,7 @@ def new_with_slugs?
     def persisted_with_slug_changes?
       if localized?
         changes = _slugs_change
-        return (persisted? && false) if changes.nil?
+        return false if changes.nil?
```
Comment (johnnyshields) [inline lib/mongoid/slug.rb:310]: This was an unrelated rubocop failure

**Comment by johnnyshields** [comment]
@mikekosulin great work, merged! Thanks.

**Pull request merged**